rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java

```diff
@@ -2537,38 +2535,69 @@ private int determineParenthesisLevel(int childLineNumber, int parentLineNumber,
         int start = sourceLineNumberOffsets[startingLineNumber - 1] + startingColumn - 1;
         int end = sourceLineNumberOffsets[endingLineNumber - 1] + endingColumn - 1;
 
-        // Determine level of parentheses by going through the source
+        // Determine level of parentheses by going through the source, don't count parentheses in comments, string literals and regexes
         int count = 0;
+        Delimiter delimiter = null;
```

Comment:
Is it safe to assume that we can "ignore" the parens in G-strings like: `"Hello ${(to).you}!"`?

Reply:
Interesting question! Answer is "yes", because the placeholder expressions are parsed as new groovy elements. To prove my point, I added following test:

```java
@Test
void gStringWithStringLiteralsWithParentheses() {
    rewriteRun(
      groovy(
        """
         "Hello ${from(":-)").via(''':-|(''').via(":-)").to(':-(')}!"
         """
      )
    );
}
```

💪🏻.